### PR TITLE
obb clipping

### PR DESCRIPTION
**Pull request by kfarr**
barely working v1

enable by pasting into console after loading a scene into editor with google3d tiles layer:
```
document.querySelector('#clipping-box').setAttribute('obb-collider2', "clipPlaneFunction: obb;"); 
document.querySelector('#clipping-box').play();
```

video:
https://github.com/user-attachments/assets/b29a31aa-d312-42b0-b87a-15988f23d775

some things this needs:
- see the other branch on how make the component usable as a boolean option in street geo component, instead of copy/paste console code --> when street-geo sets clipping ON, then the box is created with this component on it
- rename to 'obb-clipping' (component and system) 
- try setting the "known good" clipping box to be the same dimensions as the default street in the scene
- try moving the box to a child layer and/or adding obb-collider2 component to the underground box entity itself
- try moving obb-collider2 component to the 3dtiles itself using `trackedObject3D` property to reference the box
- try removing top/bottom planes to have infinite top/bottom
- once we figure out the right mechanism to allow the user to save the clipping settings, ensure it saves/loads correctly

### CODE REVIEW

File: src/components/obb-clipping.js [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+        if (boxA.intersectsOBB(boxB)) {
+          this.registerCollision(componentA, componentB);
+        }
+      }
+    }
+    this.clearCollisions();
+  }
+});
+
+AFRAME.registerComponent('obb-collider2', {
```
Comment (rahulkgupta) [inline src/components/obb-clipping.js:161]: there can be two `obb-collider2` registered?

**Review by rahulkgupta — COMMENTED** [review]
the math in this makes my head spin

File: src/components/obb-clipping.js (hunk 1 above)
Comment (rahulkgupta) [reply src/components/obb-clipping.js:161]: o one is a system

**Comment by rahulkgupta** [comment]
What's the other branch?

**Comment by kfarr** [comment]
> What's the other branch?

https://github.com/3DStreet/3dstreet/pull/877

**Comment by kfarr** [comment]
next steps:
- test if I can put the component on the street underground box instead of a root box

**Comment by kfarr** [comment]
barely working 3

issues to fix:
- data-no-pause doesn't seem to work right when loading the box at first. need to enter viewer mode and then return to 
- overloading dirt / underground box with clipping seems not like desired user behavior
- it may make sense to have an explicit clipping box that is created for this purpose at the same level as the underground box but a separate box
- it is invisible by default, shown with a wireframe if visible

**Comment by kfarr** [comment]
ok great progress so far:
- now you can specify geo, place a scene, and then use the "underground" layer to define your clipping area, then uncheck the visibility checkbox from underground

things to fix:
- improve logic to specify a target clipping element (the element to be clipped), right now '#google3d' id selector is hardcode
- improve logic to specify a source clipping element (the element to define clipping volume) ... such as if a street is not present, create a clipping box in the user root?

**Comment by kfarr** [comment]
Progress notes:
- now this will look for a target of [street] and return bounding box of the first street and subelements it finds
- todo: consider throttle ticking, remove logging, consider removing things from tick as much as possible since this is no longer a collision component

Not working yet:
- if clipping is turned on, and then i turn it off during the session, it will not remove clipping and the tick appears to still run even though the component has been removed

**Comment by kfarr** [comment]
Now ready for testing / code review:
* login with geospatial pro account
* Choose a location to enable google3d map source
* Add a street
* Click geospatial layer, open right panel, click advanced, click enableClipping
* Need more clipping? Change the size of our scene by placing objects higher or to the side. You can also place an object and toggle its visibility off.

**Comment by rahulkgupta** [comment]
works!
- testing placing things higher also works. note, scaling a lightpole doesn't change the clipping box

File: src/components/obb-clipping.js [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+    },
+    clippingDestinationSelectorString: {
+      // empty string means use the same element as the source
+      type: 'string'
+    },
+    minimumColliderDimension: { default: 0.02 }
+  },
+
+  init: function () {
+    this.tick = AFRAME.utils.throttleTick(this.tick, 250, this);
```
Comment (rahulkgupta) [inline src/components/obb-clipping.js:17]: for my education, what's the reason behind this?

File: src/editor/components/components/AddLayerPanel/createLayerFunctions.js [hunk 3]
```diff
@@ -61,7 +62,7 @@ export function createStreetmixStreet(position, streetmixURL, hideBuildings) {
   // position the street further from the current one so as not to overlap each other
   if (streetmixURL && streetmixURL !== '') {
     const definition = {
-      id: streetmixURL,
+      id: createUniqueId(),
```
Comment (rahulkgupta) [inline src/editor/components/components/AddLayerPanel/createLayerFunctions.js:65]: why this change?

File: src/components/street-geo.js [hunk 4]
```diff
@@ -113,6 +124,10 @@ AFRAME.registerComponent('street-geo', {
     const create3DtilesElement = () => {
       const google3dElement = document.createElement('a-entity');
       google3dElement.setAttribute('data-no-pause', '');
+      google3dElement.id = 'google3d';
+      if (data.enableClipping) {
+        google3dElement.setAttribute('obb-clipping', '');
```
Comment (rahulkgupta) [inline src/components/street-geo.js:129]: just to confirm, setting this attribute is what enables the `obb-clipping` component to "run" right?

**Review by rahulkgupta — COMMENTED** [review]
mostly just questions that dont need an immediate answer

File: src/components/obb-clipping.js (hunk 2 above)
Comment (kfarr) [reply src/components/obb-clipping.js:17]: this throttles tick to only fire every 250ms instead of every frame. recalculating the bounding box and then traversing all the tiles to apply the clipping planes is very expensive.

better solution is to monitor for pose changes on target object OR tiles loaded events from 3dtiles runtime object via threejs-loader-tiles library and ONLY THEN initiate updating obb and clipping planes, but throttling is a cheap easy way to do it for now

File: src/components/street-geo.js (hunk 4 above)
Comment (kfarr) [reply src/components/street-geo.js:129]: yes this adds the obb-clipping component to the 3dtiles which starts it running

File: src/editor/components/components/AddLayerPanel/createLayerFunctions.js (hunk 3 above)
Comment (kfarr) [reply src/editor/components/components/AddLayerPanel/createLayerFunctions.js:65]: streetmixURL is not unique, dragging and dropping the same street preset from the add layer panel will cause id collision errors

File: src/components/obb-clipping.js [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+      // Update OBB visual representation.
+      if (renderColliderMesh) {
+        renderColliderMesh.matrixWorld.copy(auxMatrix);
+      }
+
+      // Reset OBB with AABB and apply entity matrix. applyMatrix4 changes OBB internal state.
+      obb.copy(this.aabb);
+      obb.applyMatrix4(auxMatrix);
+
+      const clipPlanes = this.createPlanesFromOBB(obb);
```
Comment (dmarcos) [inline src/components/obb-clipping.js:397]: you don't need to update the clipping planes every frame. Can do similar to bounding box recalculation. Consider scenarios when you want to recalculate: rotation, scale, position changes

Comment (dmarcos) [reply src/components/obb-clipping.js:397]:
```
    const clipPlanes = this.createPlanesFromOBB(obb);
    this.applyClippingPlanes(clipPlanes);
```
